libr/bin/format/elf/elf.c

```diff
@@ -368,134 +372,368 @@ static RBinElfSection* get_section_by_name(struct Elf_(r_bin_elf_obj_t) *bin, co
 		int j;
-		IFDBG eprintf ("  %03x:", i);
+		int check_def;
+		char key[32] = {0};
+		Sdb *sdb_entry = sdb_new0 ();
+		snprintf (key, sizeof (key), "entry%d", i / 4);
+		sdb_ns_set (sdb, key, sdb_entry);
+		sdb_num_set (sdb_entry, "idx", i, 0);
+
 		for (j = 0; (j < 4) && (i + j) < num_entries; ++j) {
-			if (data[i + j] == 0) {
-				IFDBG eprintf ("   0 (*local*)    ");
-			} else if (data[i + j] == 1) {
-				IFDBG eprintf ("   1 (*global*)    ");
-			} else {
-				ut16 *d = (ut16*) (data + i + j);
-				//eprintf ("%4x%c", data[i + j] & 0x7FFF, data[i + j] & 0x8000 ? 'h' : ' ');
-				IFDBG eprintf ("%4x%c", *d & 0x7FFF, *d & 0x8000 ? 'h' : ' ');
+			char tmp_val[64] = {0};
+			snprintf (key, sizeof (key), "value%d", j);
+			switch (data[i + j]) {
+			case 0:
+				sdb_set (sdb_entry, key, "0 (*local*)", 0);
+				break;
+			case 1:
+				sdb_set (sdb_entry, key, "1 (*global*)", 0);
+				break;
+			default:
+				snprintf (tmp_val, sizeof (tmp_val), "%x ", data[i+j] & 0x7FFF);
+				check_def = true;
+
+				if (bin->version_info[DT_VERSIONTAGIDX (DT_VERNEED)]) {
+					Elf_(Verneed) vn;
+					ut64 offset;
+
+					offset = Elf_(r_bin_elf_v2p) (bin, bin->version_info[DT_VERSIONTAGIDX (DT_VERNEED)]);
+					do {
+						Elf_(Vernaux) vna;
+						ut64 a_off;
+
+						if (!r_buf_read_at (bin->b, offset, (ut8*)&vn, sizeof (vn))) {
+							eprintf ("Warning: Cannot read Verneed for Versym\n");
+							goto beach;
+						}
+
+						a_off = offset + vn.vn_aux;
+
+						do {
+							if (!r_buf_read_at (bin->b, a_off, (ut8*)&vna, sizeof (vna))) {
+								eprintf ("Warning: Cannot read Vernaux for Versym\n");
+								goto beach;
+							}
+
+							a_off += vna.vna_next;
+						} while (vna.vna_other != data[i + j] && vna.vna_next != 0);
+
+						if (vna.vna_other == data[i + j]) {
+							char val[64] = {0};
+							snprintf (val, sizeof (val), "%s(%s)", tmp_val, bin->strtab + vna.vna_name);
+							sdb_set (sdb_entry, key, val, 0);
+							check_def = false;
+							break;
+						}
+
+						offset += vn.vn_next;
+					} while (vn.vn_next);
+				}
+
+				ut64 vinfoaddr = bin->version_info[DT_VERSIONTAGIDX (DT_VERDEF)];
+
+				if (check_def && data[i + j] != 0x8001 && vinfoaddr) {
+					Elf_(Verdef) vd;
+					ut64 offset;
+
+					offset = Elf_(r_bin_elf_v2p) (bin, vinfoaddr);
+
+					do {
+						if (!r_buf_read_at (bin->b, offset, (ut8*)&vd, sizeof (vd))) {
+							eprintf ("Warning: Cannot read Verdef for Versym\n");
+							goto beach;
+						}
+
+						offset += vd.vd_next;
+					} while (vd.vd_ndx != (data[i + j] & 0x7FFF) && vd.vd_next != 0);
+
+					if (vd.vd_ndx == (data[i + j] & 0x7FFF)) {
+						Elf_(Verdaux) vda;
+						ut64 off_vda = offset - vd.vd_next + vd.vd_aux;
+						char val[64] = {0};
+
+						if (!(r_buf_read_at (bin->b, off_vda, (ut8*)&vda, sizeof (vda)))) {
+							eprintf ("Warning: Cannot read Verdaux for Versym\n");
+							goto beach;
+						}
+
+						const char *name = bin->strtab + vda.vda_name;
```

Comment:
do not define variables in the middle of nowhere, you can just use sdb_fmt and avoid defining that variable that it's used once